File: src/Hello/Hello.rb
#!/usr/bin/env ruby

require 'TripsModule/trips_module'

class Hello
  include TripsModule

  def initialize(argv)
    super
    @name = :Hello
    init()
    add_handler(KQML.from_s('(request &key :content (hello . *))'),
		method(:handle_hello))
  end

  def handle_hello(msg)
    send_msg(KQML[:tell, content:
      [:hello, *(msg.key?(:sender) ? [msg[:sender]] : [])]
    ] + KQML.in_reply_to(msg))
  end
end

Hello.new(ARGV).run() if ($0 == __FILE__)

File: src/Hello/Hello.pl
#!/usr/bin/env perl

package Hello;
use TripsModule::TripsModule;
@ISA = qw(TripsModule);

use strict vars;

sub init {
  my $self = shift;
  $self->{name} = 'Hello';
  $self->SUPER::init();
  $self->send_msg('(subscribe :content (request &key :content (hello . *)))');
}

sub receive_request {
  my ($self, $msg, $content) = @_;
  my $verb = lc($content->{verb});
  if ($verb eq 'hello') {
    $self->reply_to_msg($msg,
      '(tell :content (hello' .
      (exists($content->{sender}) ? ' ' . $content->{sender} : '') . '))'
    );
  } else {
    $self->reply_to_msg($msg,
      qq/(error :comment "unknown request verb $verb")/);
  }
}

Hello->new(@ARGV)->run() if ($0 eq __FILE__);

1;

